libs/wingc/src/type_check.rs

```diff
@@ -2129,38 +2185,50 @@ impl<'a> TypeChecker<'a> {
 							}
 						}
 					}
-					_ => {}
+					_ => {
+						panic!("Unexpected symbol kind: {:?} in class env", symbol)
+					}
 				}
 			}
 		}
 
 		return new_type;
 	}
 
-	fn expr_maybe_type(&mut self, expr: &Expr, env: &SymbolEnv, statement_idx: usize) -> Option<TypeRef> {
+	fn expr_maybe_type(
+		&mut self,
+		expr: &Expr,
+		env: &SymbolEnv,
+		statement_idx: usize,
+	) -> Option<(TypeRef, UserDefinedType)> {
 		// TODO: we currently don't handle parenthesized expressions correctly so something like `(MyEnum).A` or `std.(namespace.submodule).A` will return true, is this a problem?
 		let mut path = vec![];
 		let mut curr_expr = expr;
 		loop {
 			match &curr_expr.kind {
 				ExprKind::Reference(reference) => match reference {
 					Reference::Identifier(symbol) => {
-						path.push(symbol);
+						path.push(symbol.clone());
 						break;
 					}
 					Reference::NestedIdentifier { object, property } => {
-						path.push(property);
+						path.push(property.clone());
 						curr_expr = &object;
 					}
+					Reference::TypeProperty { .. } => {
+						panic!("Type property references cannot be a type name because they have a property");
```

Comment:
Can we make this into a diagnostic?

Reply:
In the current implementation this is clearly a bug. This method is used to check if we need to turn a `Reference::NestedIdentifier` into a `Reference::TypeProperty`. So we should never get here if the input expression is a `Reference::TypeProperty`. I might refactor this to make it clearer and get rid of this `panic`.
In general I'm really against adding diagnostics when we know it's a bug, because then we might never notice we have a bug in our code and treat it the same as bad input.

Reply:
Ah ok, I wasn't sure - sounds good 👍

Reply:
Added some comments and also renamed the variants in `Reference` to be clearer (I hope).